pkg/controller/gc/gc_controller.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gc
+
+import (
+	"sort"
+	"sync"
+	"time"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/client/record"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/controller"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util"
+	"k8s.io/kubernetes/pkg/watch"
+
+	"github.com/golang/glog"
+)
+
+const (
+	fullResyncPeriod = 0
+	gcCheckPeriod    = 30 * time.Second
+)
+
+type GCController struct {
+	kubeClient client.Interface
+	podControl controller.PodControlInterface
+	podStore   cache.StoreToPodLister
+	threshold  int
+}
+
+func New(kubeClient client.Interface, threshold int) *GCController {
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(glog.Infof)
+	eventBroadcaster.StartRecordingToSink(kubeClient.Events(""))
+
+	gcc := &GCController{
+		kubeClient: kubeClient,
+		podControl: controller.RealPodControl{
+			Recorder:   eventBroadcaster.NewRecorder(api.EventSource{Component: "pod-garbage-collector"}),
+			KubeClient: kubeClient,
+		},
+		threshold: threshold,
+	}
+
+	gcc.podStore.Store, _ = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return gcc.kubeClient.Pods(api.NamespaceAll).List(labels.Everything(), fields.Everything())
```

Comment:
How do we define when docker container garbage collection should kick in on the node?  Need to check.  In general, it seems like you want to define against terminal pods since those are the only viable pods to garbage collect.  I don't want you to do something just because I have larger numbers of pending or running pods per se.